libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -1006,6 +1008,152 @@ std::string YulUtilFunctions::resizeDynamicArrayFunction(ArrayType const& _type)
+						<transitLongToShort>(array, newLen)
+					}
+					default {
+						<changeLengthShort>(array, sload(array), newLen)
+					}
+				}
+			})")
+			("functionName", functionName)
+			("dataPosition", arrayDataAreaFunction(_type))
+			("partialClearStorageSlot", partialClearStorageSlotFunction())
+			("clearStorageRange", clearStorageRangeFunction(*_type.baseType()))
+			("shl", shiftLeftFunctionDynamic())
+			("transitLongToShort", byteArrayTransitLongToShortFunction(_type))
+			("changeLengthShort", changeLengthOfShortByteArrayFunction(_type))
+			.render();
+	});
+}
+
+string YulUtilFunctions::increaseByteArraySizeFunction(ArrayType const& _type)
+{
+	string functionName = "byte_array_increase_size_to_unpacked_" + _type.identifier();
+	return m_functionCollector.createFunction(functionName, [&]() {
+		return Whiskers(R"(
+			function <functionName>(array, oldLen, newLen) {
+				switch lt(oldLen, 32)
+				case 0 {
+					// in this case array stays unpacked, so we just set new length
+					sstore(array, or(<shl>(1, newLen), 1))
+				}
+				default {
+					let data := and(not(0xff), sload(array))
+					switch lt(newLen, 32)
+					case 0 {
+						// we need to copy elements to data area as we changed array from packed to unpacked
+						sstore(<dataPosition>(array), data)
+						sstore(array, or(<shl>(1, newLen), 1))
+					}
+					default {
+						// here array stays packed, we just need to increase length
+						sstore(array, or(<shl>(1, newLen), data))
+					}
+				}
+			})")
+			("functionName", functionName)
+			("dataPosition", arrayDataAreaFunction(_type))
+			("shl", shiftLeftFunctionDynamic())
+			.render();
+	});
+}
+
+string YulUtilFunctions::byteArrayTransitLongToShortFunction(ArrayType const& _type)
+{
+	string functionName = "transit_byte_array_long_to_short_" + _type.identifier();
+	return m_functionCollector.createFunction(functionName, [&]() {
+		return Whiskers(R"(
+			function <functionName>(array, len) {
+				// we need to copy elements from old array to new
+				// we want to copy only elements that are part of the array after resizing
+				let mask := <shl>(mul(8, sub(32, len)), <ones>)
+				let dataPos := <dataPosition>(array)
+				let data := and(sload(dataPos), mask)
+				sstore(array, or(data, <shl>(1, len)))
+				sstore(dataPos, 0)
+			})")
+			("functionName", functionName)
+			("dataPosition", arrayDataAreaFunction(_type))
+			("shl", shiftLeftFunctionDynamic())
+			("ones", formatNumber((bigint(1) << 256) - 1))
+			.render();
+	});
+}
+
+string YulUtilFunctions::changeLengthOfShortByteArrayFunction(ArrayType const& _type)
+{
+	string functionName = "change_length_of_short_byte_array_" + _type.identifier();
+	return m_functionCollector.createFunction(functionName, [&]() {
+		return Whiskers(R"(
+			function <functionName>(array, data, len) {
+				// we want to save only elements that are part of the array after resizing
+				// others should be set to zero
+				let mask := <shl>(mul(8, sub(32, len)), <ones>)
```

Comment:
Extract into function?

Reply:
I have deleted this function as I changed it with more usable one which just extracts usable area and sets length do data. And `sstore` is done at the place of calling that function. It looks much better now I would say